core/shared/src/main/scala/fs2/concurrent/Queue.scala

```diff
@@ -572,6 +621,113 @@ object InspectableQueue {
         }
       }
     }
+
+    private[fs2] def forStrategyNoneTerminated[F[_]: Concurrent, S, A](
+        strategy: PubSub.Strategy[Option[A], Option[Chunk[A]], S, Int]
+    )(headOf: S => Option[A])(sizeOf: S => Int): G[NoneTerminatedInspectableQueue[F, A]] = {
+      implicit val SyncG: Sync[G] = G
+
+      implicit def eqInstance: Eq[S] = Eq.fromUniversalEquals[S]
+
+      PubSub.in[G].from(PubSub.Strategy.Inspectable.strategy(strategy)).map { pubSub =>
+        new NoneTerminatedInspectableQueue[F, A] {
+
+          override def dequeueChunk(maxSize: Int): Stream[F, A] =
+            pubSub
+              .getStream(Right(maxSize))
+              .map(_.toOption.flatten)
+              .unNoneTerminate
+              .flatMap(Stream.chunk)
+
+          override def dequeueBatch: Pipe[F, Int, A] =
+            _.evalMap(sz => pubSub.get(Right(sz)).map(_.toOption.flatten)).unNoneTerminate
+              .flatMap(Stream.chunk)
+
+          override def peek1: F[A] =
+            Sync[F].bracket(Sync[F].delay(new Token)) { token =>
+              def take: F[A] =
+                pubSub.get(Left(Some(token))).flatMap {
+                  case Left(s) =>
+                    headOf(s) match {
+                      case None    => take
+                      case Some(a) => Applicative[F].pure(a)
+                    }
+
+                  case Right(chunk) =>
+                    Sync[F].raiseError(
+                      new Throwable(
+                        s"Inspectable `peek1` requires state to be returned, got: $chunk"
+                      )
+                    )
+                }
+
+              take
+            }(token => pubSub.unsubscribe(Left(Some(token))))
+
+          override def size: Stream[F, Int] =
+            Stream
+              .bracket(Sync[F].delay(new Token))(token => pubSub.unsubscribe(Left(Some(token))))
+              .flatMap { token =>
+                pubSub.getStream(Left(Some(token))).flatMap {
+                  case Left(s)  => Stream.emit(sizeOf(s))
+                  case Right(_) => Stream.empty // impossible
+                }
+              }
+
+          override def getSize: F[Int] =
+            pubSub.get(Left(None)).map {
+              case Left(s)  => sizeOf(s)
+              case Right(_) => -1
+            }
+
+          override def dequeue1: F[Option[A]] =
+            pubSub.get(Right(1)).flatMap {
+              case Left(s) =>
+                Sync[F].raiseError(
+                  new Throwable(
+                    s"Inspectable `dequeue1` requires chunk of size 1 with `A` got Left($s)"
+                  )
```

Comment:
Perhaps we should we use specialised subclasses of `Throwable` for the errors we use.